gsl/gsl_byte

```diff
@@ -106,6 +106,12 @@ constexpr IntegerType to_integer(byte b) noexcept
     return {b};
 }
 
+constexpr byte to_byte(unsigned char i) noexcept
+{
```

Comment:
I have confirmed that in Apple LLVM version 7.3.0 (clang-703.0.31) I definitely get accidental narrowing. I am used to C4244 from MS compiler. I tried simply putting your suggestion in, but it fails the unit test on the constants:
```
byte a = to_byte(0x00); // Error with is_same<T, unsigned char>
```
I have had a look at the available traits: http://en.cppreference.com/w/cpp/header/type_traits and I am not seeing anything else that will help.
I played about with overloads static asserts:
```
template <int I>
constexpr byte to_byte() noexcept
{
    static_assert(I >= 0 && I <= 255, "Value must be in range 0-255");
    return static_cast<byte>(I);
}

template <int I>
constexpr byte to_byte_v = to_byte<I>();
```
Which give the following for constants:
```
byte a = to_byte<0x00>();       // a bit odd
byte b = to_byte_v<0x00>;       // maybe better, still nicer than:
byte c = static_cast<byte>(0x00);
```
But I think they are a bit weird. I will think about it some more and see if there is anything I can come up with.